File: master/conf.cc
/**
 * @file conf.cc
 * @brief conf class implementation
 */

#include <iostream>
#include <fstream>
#include <istream>
#include <sstream>
#include <algorithm>
#include <iterator>

#include <string>
#include <list>
#include <vector>

#include "global.h"
#include "utils.h"

#include "conf.h"
#include "sos.h"

/**
 * @brief Parses the default configuration file
 */

bool conf::parse (void)
{
    file_ = CONF_PATH ;
    return conf::parse_file () ;
}

/**
 * @brief Parses the given configuration file
 */

bool conf::parse (const std::string &file)
{
    file_ = file ;
    return conf::parse_file () ;
}

/**
 * @brief Configuration file dumper
 *
 * This operator dumps the intermediate representation of the
 * configuration file to the output stream (as a simili configuration
 * file).
 */

std::ostream& operator<< (std::ostream &os, const conf &cf)
{
    if (cf.done_)
    {
	for (auto &h : cf.httplist_)
	    os << "http-server"
		<< " listen " << h.listen
	    	<< " port " << h.port
		<< " threads " << h.threads
		<< "\n" ;
	for (auto &n : cf.nslist_)
	{
	    os << "namespace "
		<< (n.type == conf::NS_ADMIN ? "admin" :
		    (n.type == conf::NS_SOS ? "sos" :
		    (n.type == conf::NS_WELL_KNOWN ?  "well-known" :
			"(unknown)")))
		<< " " ;
	    if (n.prefix.size () == 0)
		os << '/' ;
	    for (auto &p : n.prefix)
	    	os << '/' << p ;
	    os << "\n" ;
	}
	for (int i = 0 ; i < NTAB (cf.timers) ; i++)
	{
	    const char *p ;
	    switch (i)
	    {
		case conf::I_FIRST_HELLO :
		    p = "firsthello" ;
		    break ;
		case conf::I_INTERVAL_HELLO :
		    p = "hello" ;
		    break ;
		case conf::I_SLAVE_TTL :
		    p = "slavettl" ;
		    break ;
	    }
	    os << "timer " << p << " " << cf.timers [i] << "\n" ;
	}
	for (auto &n : cf.netlist_)
	{
	    os << "network " ;
	    switch (n.type)
	    {
		case conf::NET_ETH :
		    os << "ethernet " << n.net_eth.iface
			<< " ethertype 0x"
				<< std::hex << n.net_eth.ethertype << std::dec ;
		    break ;
		case conf::NET_802154 :
		    os << "802.15.4" << n.net_eth.iface
			<< " type " << (n.net_802154.type == conf::NET_802154_XBEE ? "xbee" : "(none)")
			<< " addr " << n.net_802154.addr
			<< " panid " << n.net_802154.panid
			<< " channel " << n.net_802154.channel
			;
		    break ;
		default :
		    os << "(unrecognized network)\n" ;
		    break ;
	    }
	    os << " mtu " << n.mtu << "\n" ;
	}
	for (auto &s : cf.slavelist_)
	    os << "slave id " << s.id
		<< " ttl " << s.ttl
		<< " mtu " << s.mtu
		<< "\n" ;
    }
    return os ;
}

/**
 * @brief Configuration file dumper
 *
 * This operator dumps the intermediate representation of the
 * configuration file to the output stream (as a simili configuration
 * file in HTML).
 */

std::string conf::html_debug (void)
{
    std::ostringstream oss ;

    oss << *this ;
    return oss.str () ;
}

/******************************************************************************
 * Configuration file parser
 */

#define	HELP_ALL	0
#define	HELP_HTTP	1
#define	HELP_NAMESPACE	2
#define	HELP_TIMER	3
#define	HELP_NETWORK	4
#define	HELP_SLAVE	5
#define	HELP_NETETH	(HELP_SLAVE+1)
#define	HELP_NET802154	(HELP_NETETH+1)

static const char *syntax_help [] =
{
    "http-server, namespace, timer, network, or slave",

    "http-server [listen <addr>] [port <num>] [threads <num>]",
    "namespace <admin|sos|well-known> <path>",
    "timer <firsthello|hello|slavettl|http> <value in s>",
    "network <ethernet|802.15.4> ...",
    "slave id <id> [ttl <timeout in s>] [mtu <bytes>]",

    "network ethernet <iface> [mtu <bytes>] [ethertype [0x]<val>]",
    "network 802.15.4 <iface> type <xbee> addr <addr> panid <id> [channel <chan>] [mtu <bytes>]",
} ;

bool conf::parse_file (void)
{
    std::ifstream f (file_) ;
    std::string line ;
    bool r ;

    r = true ;				// everything went well
    if (f.is_open ())
    {
	lineno_ = 0 ;

	while (getline (f, line))
	{
	    lineno_++ ;
	    r = conf::parse_line (line) ;
	    if (! r)
		break ;
	}
	if (f.bad ())
	    r = false ;
	f.close () ;
    }
    else r = false ;

    if (r)
	done_ = true ;

    return r ;
}

bool conf::parse_line (std::string &line)
{
    std::vector <std::string> tokens ;

    /*
     * First pass: remove comments
     */

    std::string::size_type sharp ;

    sharp = line.find ('#') ;		// search starting from position 0
    if (sharp != std::string::npos)
	line.erase (sharp) ;

    /*
     * Second pass: split line into tokens
     */

    std::stringstream ss (line) ;
    std::string buf ;

    while (ss >> buf)			// read word by word
    {
	tokens.push_back (buf) ;
    }

    /*
     * Third pass: quick & dirty line parser
     */

    int asize ;				// vector size
    int i ;				// current index in vector
    int r = true ;

    asize = tokens.size () ;
    i = 0 ;

    if (i < asize)
    {
	if (tokens [i] == "http-server")
	{
	    cf_http c ;

	    c.listen = "" ;
	    c.port = "" ;
	    c.threads = 0 ;

	    i++ ;
	    for ( ; i + 1 < asize ; i += 2)
	    {
		if (tokens [i] == "listen")
		{
		    if (c.listen != "")
		    {
			parse_error_dup_token (tokens [i], HELP_HTTP) ;
			r = false ;
			break ;
		    }
		    else c.listen = tokens [i+1] ;
		}
		else if (tokens [i] == "port")
		{
		    if (c.port != "")
		    {
			parse_error_dup_token (tokens [i], HELP_HTTP) ;
			r = false ;
			break ;
		    }
		    else c.port = tokens [i+1] ;
		}
		else if (tokens [i] == "threads")
		{
		    if (c.threads != 0)
		    {
			parse_error_dup_token (tokens [i], HELP_HTTP) ;
			r = false ;
			break ;
		    }
		    else c.threads = std::stoi (tokens [i+1]) ;
		}
		else
		{
		    parse_error_unk_token (tokens [i], HELP_HTTP) ;
		    r = false ;
		    break ;
		}
	    }
	    if (r)
	    {
		if (i != asize)		// odd number of parameters
		{
		    parse_error_num_token (asize, HELP_HTTP) ;
		    r = false ;
		}
		else httplist_.push_back (c) ;
	    }
	}
	else if (tokens [i] == "namespace")
	{
	    cf_namespace c ;

	    c.type = NS_NONE ;

	    // "namespace <admin|sos> <path>",

	    i++ ;
	    if (i + 2 != asize)
	    {
		parse_error_num_token (asize, HELP_NAMESPACE) ;
		r = false ;
	    }
	    else
	    {
		c.prefix = sos::split_path (tokens [i+1]) ;
		if (tokens [i] == "admin")
		    c.type = NS_ADMIN ;
		else if (tokens [i] == "sos")
		    c.type = NS_SOS ;
		else if (tokens [i] == "well-known")
		    c.type = NS_WELL_KNOWN ;
		else
		{
		    parse_error_unk_token (tokens [i], HELP_NAMESPACE) ;
		    r = false ;
		}
		if (r)
		    nslist_.push_back (c) ;
	    }
	}
	else if (tokens [i] == "timer")
	{
	    i++ ;

	    if (i + 2 != asize)
	    {
		parse_error_num_token (asize, HELP_TIMER) ;
		r = false ;
	    }
	    else
	    {
		int idx ;

		if (tokens [i] == "firsthello")
		    idx = I_FIRST_HELLO ;
		else if (tokens [i] == "hello")
		    idx = I_INTERVAL_HELLO ;
		else if (tokens [i] == "slavettl")
		    idx = I_SLAVE_TTL ;
		else
		    idx = -1 ;

		if (idx == -1)
		{
		    parse_error_unk_token (tokens [i], HELP_TIMER) ;
		    r = false ;
		}
		else if (timers [idx] != 0)
		{
		    i++ ;
		    timers [idx] = std::stoi (tokens [i]) ;
		}
		else
		{
		    parse_error_dup_token (tokens [0], HELP_TIMER) ;
		    r = false ;
		}
	    }
	}
	else if (tokens [i] == "network")
	{
	    cf_network c ;

	    c.type = NET_NONE ;
	    c.mtu = 0 ;

	    i++ ;
	    if (i < asize)
	    {
		if (tokens [i] == "ethernet")
		{
		    c.type = NET_ETH ;
		    i++ ;

		    if (i < asize)
		    {
			c.net_eth.iface = tokens [i] ;
			i++ ;

			for ( ; i + 1 < asize ; i += 2)
			{
			    if (tokens [i] == "mtu")
			    {
				if (c.mtu != 0)
				{
				    parse_error_dup_token (tokens [i], HELP_NETETH) ;
				    r = false ;
				    break ;
				}
				else c.mtu = std::stoi (tokens [i+1]) ;
			    }
			    else if (tokens [i] == "ethertype")
			    {
				if (c.net_eth.ethertype != 0)
				{
				    parse_error_dup_token (tokens [i], HELP_NETETH) ;
				    r = false ;
				    break ;
				}
				else
				{
				    std::istringstream is (tokens [i+1]) ;

				    if (tokens [i+1].substr (0, 2) == "0x")
					is >> std::hex >> c.net_eth.ethertype ;
				    else
					is >> c.net_eth.ethertype ;
				}
			    }
			    else
			    {
				parse_error_unk_token (tokens [i], HELP_NETETH) ;
				r = false ;
				break ;
			    }
			}
		    }
		    else
		    {
			parse_error_num_token (asize, HELP_NETETH) ;
			r = false ;
		    }

		    if (r)
		    {
			if (i != asize)		// odd number of parameters
			{
			    parse_error_num_token (asize, HELP_NETETH) ;
			    r = false ;
			}
		    }
		}
		else if (tokens [i] == "802.15.4")
		{
		    c.type = NET_802154 ;
		    c.net_802154.type = NET_802154_NONE ;
		    c.net_802154.addr = "" ;
		    c.net_802154.panid = "" ;
		    c.net_802154.channel = 0 ;
		    i++ ;

		    if (i < asize)
		    {
			c.net_802154.iface = tokens [i] ;
			i++ ;

			for ( ; i + 1 < asize ; i += 2)
			{
			    if (tokens [i] == "mtu")
			    {
				if (c.mtu != 0)
				{
				    parse_error_dup_token (tokens [i], HELP_NET802154) ;
				    r = false ;
				    break ;
				}
				else c.mtu = std::stoi (tokens [i+1]) ;
			    }
			    else if (tokens [i] == "type")
			    {
				if (c.net_802154.type != NET_802154_NONE)
				{
				    parse_error_dup_token (tokens [i], HELP_NET802154) ;
				    r = false ;
				    break ;
				}
				else
				{
				    if (tokens [i+1] == "xbee")
					c.net_802154.type = NET_802154_XBEE ;
				    else
				    {
					parse_error_unk_token (tokens [i+1], HELP_NET802154) ;
					r = false ;
					break ;
				    }
				}
			    }
			    else if (tokens [i] == "addr")
			    {
				if (c.net_802154.addr != "")
				{
				    parse_error_dup_token (tokens [i], HELP_NET802154) ;
				    r = false ;
				    break ;
				}
				else c.net_802154.addr = tokens [i+1] ;
			    }
			    else if (tokens [i] == "panid")
			    {
				if (c.net_802154.panid != "")
				{
				    parse_error_dup_token (tokens [i], HELP_NET802154) ;
				    r = false ;
				    break ;
				}
				else c.net_802154.panid = tokens [i+1] ;
			    }
			    else if (tokens [i] == "channel")
			    {
				if (c.net_802154.channel != 0)
				{
				    parse_error_dup_token (tokens [i], HELP_NET802154) ;
				    r = false ;
				    break ;
				}
				else c.net_802154.channel = std::stoi (tokens [i+1]) ;
			    }
			    else
			    {
				parse_error_unk_token (tokens [i], HELP_NET802154) ;
				r = false ;
				break ;
			    }
			}
		    }
		    else
		    {
			parse_error_num_token (asize, HELP_NET802154) ;
			r = false ;
		    }

		    if (r)
		    {
			if (i != asize)		// odd number of parameters
			{
			    parse_error_num_token (asize, HELP_NET802154) ;
			    r = false ;
			}
		    }
		}
		else
		{
		    parse_error_unk_token (tokens [i], HELP_NETWORK) ;
		    r = false ;
		}

		if (r)
		    netlist_.push_back (c) ;
	    }
	    else parse_error_mis_token (HELP_NETWORK) ;
	}
	else if (tokens [i] == "slave")
	{
	    cf_slave c ;

	    c.id = 0 ;
	    c.ttl = 0 ;
	    c.mtu = 0 ;

	    i++ ;
	    for ( ; i + 1 < asize ; i += 2)
	    {
		if (tokens [i] == "id")
		{
		    if (c.id != 0)
		    {
			parse_error_dup_token (tokens [i], HELP_SLAVE) ;
			r = false ;
			break ;
		    }
		    else c.id = std::stoi (tokens [i+1]) ;
		}
		else if (tokens [i] == "ttl")
		{
		    if (c.ttl != 0)
		    {
			parse_error_dup_token (tokens [i], HELP_SLAVE) ;
			r = false ;
			break ;
		    }
		    else c.ttl = std::stoi (tokens [i+1]) ;
		}
		else if (tokens [i] == "mtu")
		{
		    if (c.mtu != 0)
		    {
			parse_error_dup_token (tokens [i], HELP_SLAVE) ;
			r = false ;
			break ;
		    }
		    else c.mtu = std::stoi (tokens [i+1]) ;
		}
		else
		{
		    parse_error_unk_token (tokens [i], HELP_SLAVE) ;
		    r = false ;
		    break ;
		}
	    }
	    if (r)
	    {
		if (i != asize)		// odd number of parameters
		{
		    parse_error_num_token (asize, HELP_SLAVE) ;
		    r = false ;
		}
		else slavelist_.push_back (c) ;
	    }
	}
	else
	{
	    parse_error_unk_token (tokens [i], HELP_ALL) ;
	    r = false ;
	}
    }

    /*
     * Fourth pass: set default values
     */

    for (auto &h : httplist_)
    {
	if (h.port == "")
	    h.port = DEFAULT_HTTP_PORT ;
	if (h.threads == 0)
	    h.threads = DEFAULT_HTTP_THREADS ;
	if (h.listen == "")
	    h.listen = DEFAULT_HTTP_LISTEN ;
    }

    if (timers [I_FIRST_HELLO] == 0)
	timers [I_FIRST_HELLO] = DEFAULT_FIRST_HELLO ;
    if (timers [I_INTERVAL_HELLO] == 0)
	timers [I_INTERVAL_HELLO] = DEFAULT_INTERVAL_HELLO ;
    if (timers [I_SLAVE_TTL] == 0)
	timers [I_SLAVE_TTL] = DEFAULT_SLAVE_TTL ;

    for (auto &s : slavelist_)
	if (s.ttl == 0)
	    s.ttl = timers [I_SLAVE_TTL] ;

    for (auto &n : netlist_)
    {
	if (n.type == NET_ETH && n.net_eth.ethertype == 0)
	    n.net_eth.ethertype = ETHTYPE_SOS ;
	if (n.type == NET_802154 && n.net_802154.channel == 0)
	    n.net_802154.channel = DEFAULT_802154_CHANNEL ;
    }

    if (r)
	done_ = true ;

    return r ;
}

void conf::parse_error (const std::string msg, int help)
{
    std::cerr << file_ << "(" << lineno_ << "): " << msg << "\n" ;
    if (help >= 0)
	std::cerr << "\tusage: " << syntax_help [help] << "\n" ;
}

void conf::parse_error_num_token (const int n, int help)
{
    std::string msg ;

    msg = "invalid number of tokens (" + std::to_string (n) + ")" ;
    parse_error (msg, help) ;
}

void conf::parse_error_unk_token (const std::string tok, int help)
{
    std::string msg ;

    msg = "unrecognized token '" + tok + "'" ;
    parse_error (msg, help) ;
}

void conf::parse_error_dup_token (const std::string tok, int help)
{
    std::string msg ;

    msg = "duplicate token '" + tok + "'" ;
    parse_error (msg, help) ;
}

void conf::parse_error_mis_token (int help)
{
    std::string msg ;

    msg = "missing token/value" ;
    parse_error (msg, help) ;
}

File: pySos/conf.py
from enum import Enum, IntEnum
from io import StringIO

class Conf:
    """
    Configuration file handling
    The conf class is used to read the configuration file and
    represent the configuration through an intermediate form,
     which is composed of instance variables that the master
     class can use.
    """
    # Define the constants defined in conf.h and help strings
    DEFAULT_FIRST_HELLO = 3
    DEFAULT_INTERVAL_HELLO = 10
    DEFAULT_SLAVE_TTL = 3600
    DEFAULT_HTTP_PORT = 'http'
    DEFAULT_HTTP_LISTEN = '*'
    DEFAULT_HTTP_THREADS = 5
    DEFAULT_154_CHANNEL = 12
    syntax_help = {'ALL': 'http-server, namespace, timer, network, or slave',
                   'HTTP': 'http-server [listen <addr>] [port <num>] [threads <num>]',
                   'NAMESPACE': 'namespace <admin|sos|well-known> <path>',
                   'TIMER': 'timer <firsthello|hello|slavettl|http> <value in s>',
                   'NETWORK': 'network <ethernet|802.15.4> ...',
                   'SLAVE': 'slave id <id> [ttl <timeout in s>] [mtu <bytes>]',
                   'NETETH': 'network ethernet <iface> [mtu <bytes>] [ethertype [0x]<val>]',
                   'NET154': 'network 802.15.4 <iface> type <xbee> addr <addr> panid <id> [channel <chan>] [mtu <bytes>]'}

    # Define the enums defined in conf.h
    class CfNsType(Enum):
        NS_NONE = 0
        NS_ADMIN = 1
        NS_SOS = 2
        NS_WELL_KNOWN = 3

    class CfTimerIndex(IntEnum):
        I_FIRST_HELLO = 0
        I_INTERVAL_HELLO = 1
        I_SLAVE_TTL = 2
        I_LAST = 3

    class NetType(Enum):
        NET_NONE = 0
        NET_ETH = 1
        NET_154 = 2

    class Net154Type(Enum):
        NET_154_NONE = 0
        NET_154_XBEE = 1

    # Define the structs defined in conf.h
    class CfHttp:
        def __init__(self):
            self.listen = ''
            self.port = ''
            self.threads = 0

    class CfNamespace:
        def __init__(self):
            self.type = Conf.CfNsType.NS_NONE

    class CfNetEth:
        def __init__(self):
            self.ethertype = 0

    class CfNet154:
        def __init__(self):
            self.type = Conf.Net154Type.NET_154_NONE
            self.addr, self.panid, self.iface = '', '', ''
            self.channel = 0

    class CfNetwork:
        def __init__(self):
            self.type, self.mtu = Conf.NetType.NET_NONE, 0

    class CfSlave:
        def __init__(self):
            self.id, self.ttl, self.mtu = 0, 0, 0

    # And finally the actual configuration functions

    def __init__(self):
        self.done = False
        self.httplist, self.nslist, self.netlist, self.slavelist = [], [], [], []
        self.timers = [0] * 3
        self.lineno = 0
        self.file = None

    def parse(self, file_="./sosd.conf"):
        return self.parse_file(file_)

    def parse_file(self, file_):
        self.file = file_
        r = False
        try:
            self.file = file_
            f = open(self.file, 'r')
            self.lineno = 0
            for line in f:
                self.lineno += 1
                r = self.parse_line(line)
                if not r:
                    break
        except OSError:
            return False
        if r:
            self.done = True
        self.file = None
        return r

    def parse_line(self, l):
        # First pass : remove comments
        if '#' in l:
            l = l[:l.find('#')]

        # Second pass : split line into tokens
        tokens = l.split()
        r = True

        # Third pass : quick & dirty line parser
        a = len(tokens)
        if a != 0:
            if tokens[0] == 'http-server':
                c = self.CfHttp()
                for i in range(1, a - 1, 2):
                    if tokens[i] == 'listen':
                        if c.listen != '':
                            self.parse_error_dup_token(tokens[i], 'HTTP')
                            r = False
                        else:
                            c.listen = tokens[i+1]
                    elif tokens[i] == 'port':
                        if c.port != '':
                            self.parse_error_dup_token(tokens[i], 'HTTP')
                            r = False
                        else:
                            c.port = tokens[i+1]
                    elif tokens[i] == 'threads':
                        if c.threads != 0:
                            self.parse_error_dup_token(tokens[i], 'HTTP')
                            r = False
                        else:
                            c.threads = int(tokens[i+1])
                    else:
                        self.parse_error_unk_token(tokens[i], 'HTTP')
                        r = False
                if r:
                    if a % 2 != 1:  # Odd number of parameters
                        self.parse_error_num_token(a, 'HTTP')
                        r = False
                    else:
                        self.httplist.append(c)
            elif tokens[0] == 'namespace':
                c = self.CfNamespace()
                if a != 3:
                    self.parse_error_num_token(a, 'NAMESPACE')
                    r = False
                else:
                    c.prefix = tokens[2].split('/')[1:]  # Remove leading empty string
                    try:
                        c.type = {'admin': self.CfNsType.NS_ADMIN,
                                  'sos': self.CfNsType.NS_SOS,
                                  'well-known': self.CfNsType.NS_WELL_KNOWN}[tokens[1]]
                    except KeyError:
                        self.parse_error_unk_token(tokens[1], 'NAMESPACE')
                        r = False
                    if r:
                        self.nslist.append(c)
            elif tokens[0] == 'timer':
                if a != 3:
                    self.parse_error_num_token(a, 'TIMER')
                    r = False
                else:
                    try:
                        idx = self.CfTimerIndex[{'firsthello': 'I_FIRST_HELLO',
                                                 'hello': 'I_INTERVAL_HELLO',
                                                 'slavettl': 'I_SLAVE_TTL'}[tokens[1]]]
                        self.timers[idx] = int(tokens[2])
                    except KeyError:
                        self.parse_error_unk_token(tokens[1], 'TIMER')
                        r = False
            elif tokens[0] == 'network':
                if a < 3:  # 3 Mandatory arguments according to sample sosd.conf
                    self.parse_error_mis_token('NETWORK')
                else:
                    c = self.CfNetwork()
                    if tokens[1] == 'ethernet':
                        # I think ethernet support is on in python, so
                        # I removed the test, for now.
                        c.type = self.NetType.NET_ETH
                        i = 2
                        if i >= a:
                            self.parse_error_num_token(a, 'NETETH')
                            r = False
                        else:
                            c.net_eth = self.CfNetEth()
                            c.net_eth.iface = tokens[i]
                            for i in range(i, a-1, 2):
                                if tokens[i] == 'mtu':
                                    if c.mtu != 0:
                                        self.parse_error_dup_token(tokens[i], 'NETETH')
                                        r = False
                                        break
                                    else:
                                        c.mtu = int(tokens[i+1])
                                elif tokens[i] == 'ethertype':
                                    if c.net_eth.ethertype != 0:
                                        self.parse_error_dup_token(tokens[i], 'NETETH')
                                        r = False
                                        break
                                    else:
                                        c.net_eth.ethertype = tokens[i+1]
                                else:
                                    self.parse_error_unk_token(tokens[i], 'NETETH')
                                    r = False
                                    break
                            if r and i+2 != a:  # Odd number of parameters
                                self.parse_error_num_token(a, 'NETETH')
                                # self.parse_error('No ethernet support')
                    elif tokens[1] == '802.15.4':
                        c.type = self.NetType.NET_154
                        c.net_154 = self.CfNet154()
                        i = 3
                        if i >= a:
                            self.parse_error_num_token(a, 'NET154')
                        else:
                            c.net_154.iface = tokens[2]
                            for i in range(i, a-1, 2):
                                if tokens[i] == 'mtu':
                                    if c.mtu != 0:
                                        self.parse_error_dup_token(tokens[i], 'NET154')
                                        r = False
                                        break
                                    else:
                                        c.mtu = int(tokens[i])
                                elif tokens[i] == 'type':
                                    if c.net_154.type is not self.Net154Type.NET_154_NONE:
                                        self.parse_error_dup_token(tokens[i], 'NET154')
                                        r = False
                                        break
                                    else:
                                        if tokens[i+1] == 'xbee':
                                            c.net_154.type = self.Net154Type.NET_154_XBEE
                                        else:
                                            self.parse_error_unk_token(tokens[i+1], 'NET154')
                                            r = False
                                            break
                                elif tokens[i] == 'addr':
                                    if c.net_154.addr != '':
                                        self.parse_error_dup_token(tokens[i], 'NET154')
                                        r = False
                                        break
                                    else:
                                        c.net_154.addr = tokens[i+1]
                                elif tokens[i] == 'panid':
                                    if c.net_154.panid != '':
                                        self.parse_error_dup_token(tokens[i], 'NET154')
                                        r = False
                                        break
                                    else:
                                        c.net_154.panid = tokens[i+1]
                                elif tokens[i] == 'channel':
                                    if c.net_154.channel != 0:
                                        self.parse_error_dup_token(tokens[i], 'NET154')
                                        r = False
                                        break
                                    else:
                                        c.net_154.channel = int(tokens[i+1])
                                else:
                                    self.parse_error_unk_token(tokens[i], 'NET154')
                                    r = False
                                    break
                        if r and i != a-2:
                            self.parse_error_num_token(a, 'NET154')
                            r = False
                    else:
                        self.parse_error_unk_token(tokens[1], 'NETWORK')
                        r = False
                    if r:
                        self.netlist.append(c)
            elif tokens[0] == 'slave':
                c = self.CfSlave()
                for i in range(1, a-1, 2):
                    if tokens[i] == 'id':
                        if c.id != 0:
                            self.parse_error_dup_token(tokens[i], 'SLAVE')
                            r = False
                            break
                        else:
                            c.id = int(tokens[i+1])
                    elif tokens[i] == 'ttl':
                        if c.ttl != 0:
                            self.parse_error_dup_token(tokens[i], 'SLAVE')
                            r = False
                            break
                        else:
                            c.ttl = int(tokens[i+1])
                    elif tokens[i] == 'mtu':
                        if c.mtu != 0:
                            self.parse_error_dup_token(tokens[i], 'SLAVE')
                            r = False
                            break
                        else:
                            c.mtu = int(tokens[i+1])
                    else:
                        self.parse_error_unk_token(tokens[i], 'SLAVE')
                        r = False
                        break
                if r:
                    if i != a-2:
                        self.parse_error_num_token(a, 'SLAVE')
                    else:
                        self.slavelist.append(c)
            else:
                self.parse_error_unk_token(tokens[0], 'ALL')
                r = False

        # 4th pass : set default values
        for h in self.httplist:
            if h.port == '':
                h.port = self.DEFAULT_HTTP_PORT
            if h.listen == '':
                h.listen = self.DEFAULT_HTTP_LISTEN
            if h.threads == 0:
                h.threads = self.DEFAULT_HTTP_THREADS

        if self.timers[self.CfTimerIndex.I_FIRST_HELLO] == 0:
            self.timers[self.CfTimerIndex.I_FIRST_HELLO] = 0
        if self.timers[self.CfTimerIndex.I_INTERVAL_HELLO] == 0:
            self.timers[self.CfTimerIndex.I_INTERVAL_HELLO] = 0
        if self.timers[self.CfTimerIndex.I_SLAVE_TTL] == 0:
            self.timers[self.CfTimerIndex.I_SLAVE_TTL] = 0
                    
        for slave in self.slavelist:
            if slave.ttl == 0:
                slave.ttl = self.timers[self.CfTimerIndex.I_SLAVE_TTL]

        for n in self.netlist:
            if n.type is self.NetType.NET_ETH and n.net_eth.ethertype == 0:
                n.net_eth.ethertype = 0
            if n.type is self.NetType.NET_154 and n.net_154.channel == 0:
                n.net_154.channel = self.DEFAULT_154_CHANNEL

        if r:
            self.done = True
        return r

    def write_to_stream(self, stream):
        if self.done:
            for h in self.httplist:
                stream.write('http-server listen ' + h.listen +
                             ' port ' + h.port + 
                             ' threads ' + str(h.threads) + '\n')
            for n in self.nslist:
                s = ''
                try:
                    s = {self.CfNsType.NS_ADMIN: 'admin',
                         self.CfNsType.NS_SOS: 'sos',
                         self.CfNsType.NS_WELL_KNOWN: 'well-known'}[n.type]
                except KeyError:
                    s = '(unknown)'
                stream.write('namespace ' + s + ' ')
                if len(n.prefix) == 0:
                    stream.write('/')
                else:
                    for p in n.prefix:
                        stream.write('/' + p)
                    stream.write('\n')
            for i in range(0, len(self.timers)):
                p = {self.CfTimerIndex.I_FIRST_HELLO: 'firsthello',
                     self.CfTimerIndex.I_INTERVAL_HELLO: 'hello',
                     self.CfTimerIndex.I_SLAVE_TTL: 'slavettl' }[i]
                stream.write(p + ' ' + str(self.timers[i]) + '\n')
            for n in self.netlist:
                if n.type is self.NetType.NET_ETH:
                    stream.write('network ' + 'ethernet ' + n.net_eth.iface + 
                                 'ethertype ' + n.net_eth.ethertype)
                elif n.type is self.NetType.NET_154:
                    stream.write('network ' + '802.15.4 ' + n.net_154.iface +
                                 ' type ' + ('xbee' if n.net_154.type is self.Net154Type.NET_154_XBEE else '(none)') +
                                 ' addr ' + n.net_154.addr +
                                 ' panid ' + n.net_154.panid +
                                 ' channel ' + str(n.net_154.channel))
                else:
                    stream.write('(unrecognised network)\n')
                stream.write(' mtu ' + str(n.mtu) + '\n')
            for s in self.slavelist:
                stream.write('slave id ' + str(s.id) +
                             ' ttl ' + str(s.ttl) +
                             ' mtu ' + str(s.mtu) + '\n')

    def to_string(self):
        ss = StringIO()
        self.write_to_stream(ss)
        return ss.getvalue()

    # Error handlers
    def parse_error(self, msg, help_=None):
        import sys
        sys.stderr.write(self.file + '(' + str(self.lineno) + '):' + msg + '\n')
        if help_ is not None:
            sys.stderr.write('    usage: ' + self.syntax_help[help_]  + '\n')

    def parse_error_num_token(self, n, help_):
        self.parse_error('invalid number of tokens (' + str(n) + ')', help_)

    def parse_error_unk_token(self, tok, help_):
        self.parse_error('unrecognised token \'' + tok + '\'', help_)

    def parse_error_dup_token(self, tok, help_):
        self.parse_error('duplicate token \'' + tok + '\'', help_)

    def parse_error_mis_token(self, help_):
        self.parse_error('missing token/value', help_)

